collect_app/src/main/java/org/odk/collect/android/utilities/SelectOneWidgetUtils.java

```diff
@@ -0,0 +1,34 @@
+package org.odk.collect.android.utilities;
+
+import org.javarosa.core.model.SelectChoice;
+import org.javarosa.core.model.data.IAnswerData;
+import org.javarosa.core.model.data.SelectOneData;
+import org.javarosa.core.model.data.StringData;
+import org.javarosa.core.model.data.helper.Selection;
+import org.javarosa.form.api.FormEntryPrompt;
+
+import java.util.List;
+
+public class SelectOneWidgetUtils {
+
+    private SelectOneWidgetUtils() {
+
+    }
+
+    public static Selection getSelectedItem(FormEntryPrompt prompt, List<SelectChoice> items) {
+        IAnswerData answer = prompt.getAnswerValue();
+        if (answer == null) {
+            return null;
+        } else if (answer instanceof SelectOneData) {
+            return (Selection) answer.getValue();
+        } else if (answer instanceof StringData) { // Fast external itemset
+            for (SelectChoice item : items) {
```

Comment:
How was this done before? It'd be ideal not to loop through all items.

Reply:
We had a method in [SelectOneWidget](https://github.com/getodk/collect/pull/4242/files#diff-6ccb223484e6d1101c41b833d6cd419cab66f2b6d973907112dea076cb6b7275R82) and also it was overrode in [ItemsetWidget](https://github.com/getodk/collect/pull/4242/files#diff-03d8e0549dc87e59bb6e27487024b6b81daa3dc0152a334628da85b33b4723b3L84)

The problem is that we use DATATYPE_TEXT in case of external itemsets probably in the future we should change it adding also support for select_multiple then we won't need such checks but that's on javarosa's side.

Reply:
Both of the methods you linked to just return the string value. Is `getSelectedItem` only needed for certain appearances? The reason folks use "fast external itemsets" is because they're database-driven and therefore fast. I wouldn't want to add overhead for the common case which is a simple select one with no appearance.

Reply:
Oh sorry my mistake it's not how I wanted to explain it. The problem is that if we use fast external itemsets DATATYPE_TEXT is used which uses StringData but for normal selects we use CONTROL_SELECT_ONE which uses SelectOneData. So when we save a form with answers and open it it might receive StringData or SelectOneData and now we need to handle it since we use the same wights no matter what the source of items is.
You can notice that difference in methods I linked as well. That's why I said it would be good to use the same datatypes implementing changes in javarosa isn't that possible?

Reply:
Ah, I think I see what you mean. 

> That's why I said it would be good to use the same datatypes implementing changes in javarosa isn't that possible?

JavaRosa doesn't know anything at all about either "fast external itemsets" or `search()`. They're entirely client-side features. So no, not really. Maybe you could look at how the `search()` code ends up with `CONTROL_SELECT_ONE` types?

Reply:
The difference is that when we convert xls to xml using `search()` it's still `select1` element in xml but `select_one_external` used in xls ends up as `<input query="...` and then in Collect [if such `<input` has query](https://github.com/getodk/collect/pull/4242/files#diff-d65ca3cdb4834d0b00ca74ecde18b9930e40762a6c48e2fbb991d8e6cfd4c28dR153) we treat it as select_one question. Probably we could add such a check earlier - on javarosa's side and treat it as `CONTROL_SELECT_ONE` not `DATATYPE_TEXT`

Reply:
Hmm. I see, the body element types are different, that's right. I really don't want us to start leaking these weird features into JavaRosa where there's already enough going on. Could you please do a quick performance check before and after your change for 5k items? I think it should be really quick to generate in Excel but let me know if not. There already was a list of `SelectChoice` items generated for "fast external itemsets", right?

Reply:
I checked it using android 6 with a device which is not very fast and didn't notice any change in performance comparing this branch with the master branch. I used this form if you want to check it on your own: 
https://drive.google.com/file/d/1lXZtqhGyMJZW5BFrRyPdTcSwRgQ95wQZ/view?usp=sharing
I selected one of the last choices filling the form on purpose to make the searching longer.